vector/src/vector.rs

```diff
@@ -0,0 +1,877 @@
+                // and their entire subtrees.
+                let max_child_len = N.pow(u32::from(height));
+                let num_full_children = len / max_child_len;
+                let extra = len % max_child_len;
+                if extra > 0 {
+                    children.drop_right(num_full_children + 1);
+                    Rc::make_mut(&mut children[num_full_children]).truncate(height - 1, extra);
+                } else {
+                    children.drop_right(num_full_children);
+                }
+            }
+        }
+    }
+}
+
+/// [`Vector`] is a persistent vector (also known as a "bitmapped vector trie").
+///
+/// Most of the operations on `Vector` have similar asymptotic (but
+/// with a slower constant-factor) run-time to the same operations on the
+/// standard-library `Vec`. For example, you can quickly push an element to
+/// the back, or pop one from the back; random-access indexing is also fast.
+/// On the other hand, pushing/popping from the front or insertion/deletion in
+/// the middle are both slow. (Actually, we haven't even implemented these slow
+/// operations because we don't need them.)
+///
+/// The main advantage that [`Vector`] has over [`std::vec::Vec`] is *persistence*:
+/// you can cheaply clone a `Vector` and then modify the clone. The clone and the
+/// original will share most of their storage.
+///
+/// There's a good explanation of the data structure
+/// [here](https://hypirion.com/musings/understanding-persistent-vector-pt-1).
+///
+/// ## Branching factor `N`.
+///
+/// `Vector` stores its data in a tree, and you get to choose the branching factor
+/// `N`. It must be a power of 2 between 2 and 128 (inclusive), and something like
+/// 32 seems to be reasonable. There is a trade-off involved, of course. Large
+/// values of `N` make traditional array operations -- like random access, pushing,
+/// and popping -- fast because the tree becomes very flat. On the other hand, the
+/// copy-on-write operations suffer when `N` is large because blocks of size `N`
+/// are the smallest units of shareable data.
+///
+/// For example, if `N` is 128 and you have a vector of less than 128 entries,
+/// it will be stored in a single flat array of capacity 128. If you then clone
+/// the `Vector` and modify the clone, the whole length-128 array will be
+/// cloned. If the vector is stored as a taller tree, only the blocks on the
+/// path from the root to the modified entry will be duplicated. That is, modifying
+/// an entry in a `Vector` of length `n` and branching factor `N` has
+/// cost `O(N * log_N n)`.
+///
+/// ## Comparison to `rpds`
+///
+/// The same structure is implemented in [rpds](https://crates.io/crates/rpds),
+/// but our implementation is faster for Nickel's use-cases:
+/// - rpds's internal nodes are implemented with `Vec`, meaning that there's a
+///   double pointer indirection. We store our internal nodes inline.
```

Comment:
I don't really know their reason, but
- using vec allows their branching parameter to be a runtime parameter instead of a const generic (which are a bit annoying to use and probably weren't even available when they started)
- using arrays means using unsafe code and uninitialized stuff (I guess technically `Option` would be an ... option). We delegate this to the `imbl_sized_chunks` crate